arbor/fvm_layout.cpp

```diff
@@ -59,6 +70,73 @@ namespace {
     };
 
     ARB_DEFINE_LEXICOGRAPHIC_ORDERING(cv_param,(a.cv,a.params,a.target),(b.cv,b.params,b.target))
+
+    template <typename V>
+    optional<V> operator|(const optional<V>& a, const optional<V>& b) {
+        return a? a: b;
+    }
+
+    // For each segment given by the provided sorted sequence of segment
+    // indices, call `action` for each CV intersecting the segment, starting
+    // from the most proximal.
+    //
+    // By the ordering of CVs and segments in the discretization, with the
+    // exception of the most proximal CV in a segment, each CV will be visited
+    // once, and the visited CVs will be in increasing order. The most proximal
+    // CV (the 'parent' CV) may be visited multiple times.
+    //
+    // Action is a functional that takes the following arguments:
+    //
+    //     size_type  cv_index   The index into the total (sorted) list of
+    //                           CVs that constitute the segments.
+    //
+    //     index_type cv         The CV number (within the discretization).
+    //
+    //     value_type cv_area    The area of the CV that lies within the
+    //                           current segment.
+    //
+    //     index_type seg_index  The index into the provided sequence of
+    //                           the provided segment_indices.
+    //
+    //     index_type seg        The segment currently being iterated over.
+
+    template <typename Seq, typename Action>
+    void for_each_cv_in_segments(const fvm_discretization& D, const Seq& segment_indices, const Action& action) {
+        using index_type = fvm_index_type;
+        using size_type = fvm_size_type;
+
+        std::unordered_map<index_type, size_type> parent_cv_indices;
+        size_type cv_index = 0;
+
+        index_type seg_index = 0;
+        for (const auto& seg: segment_indices) {
+            const segment_info& seg_info = D.segments[seg];
+
+            if (seg_info.has_parent()) {
+                index_type cv = seg_info.parent_cv;
+
+                size_type i = parent_cv_indices.insert({cv, cv_index}).first->second;
+                if (i==cv_index) {
```

Comment:
Are we just checking that insertion is successful here? Why not just check `parent_cv_indices.insert({cv, cv_index}).second`?

Reply:
In the instance that we've already seen this CV, we want to retrieve the previously saved index value for passing on to the action; by grabbing the value from the first field (the iterator), we can avoid a second look-up.